src/localization/database/cost_matrix.cpp

```diff
@@ -0,0 +1,94 @@
+#include "cost_matrix.h"
+#include <fstream>
+
+#include "database/list_dir.h"
+#include "features/feature_factory.h"
+#include "localization_protos.pb.h"
+
+#include <glog/logging.h>
+
+CostMatrix::CostMatrix(const std::string &costMatrixFile) {
+  CHECK(!costMatrixFile.empty()) << "Cost matrix file is not set";
+  loadFromProto(costMatrixFile);
+}
+CostMatrix::CostMatrix(const std::string &queryFeaturesDir,
+                       const std::string &refFeaturesDir,
+                       const FeatureType &type) {
+
+  std::vector<std::string> queryFeatureFiles =
+      listProtoDir(queryFeaturesDir, ".Feature");
+  std::vector<std::string> refFeaturesFiles =
+      listProtoDir(refFeaturesDir, ".Feature");
+
+  costs_.reserve(queryFeatureFiles.size());
+  for (const auto &queryFile : queryFeatureFiles) {
+    auto queryFeature = createFeature(type, queryFile);
+    std::vector<double> row(refFeaturesFiles.size());
```

Comment:
I think you meant this here:
```cpp
std::vector<double> row{};
row.reserve(refFeaturesFiles.size());
```
Right? 🤔

Reply:
Also, do you have tests for this?

Reply:
Now I have :)